rfcs/0023-adding-annotation-field-to-tir.allocate.md

```diff
@@ -0,0 +1,201 @@
+  std::string tag;
+```
+
+To serve P1, this RFC introduces the addition of the annotations field
+
+# 3. Guide-level explanation
+
+This is not particularly a user-facing feature.
+
+
+ # 4. Reference-level explanation
+
+
+To serve P1, we propose to use :
+
+```
+class AllocateNode : public StmtNode {
+ public:
+  /*! \brief The buffer variable. */
+  Var buffer_var;
+  /*! \brief The type of the buffer. */
+  DataType dtype;
+  /*! \brief The extents of the buffer. */
+  Array<PrimExpr> extents;
+  /*! \brief Only allocate buffer when condition is satisfied. */
+  PrimExpr condition;
+  /*! \brief The body to be executed. */
+  Stmt body;
+  /*! \brief If the allocate is scoped global, this field indicates
+   *  which external memories it could be pinned to as a comma seperated
+   *  string.
+   */
++ Map<String, Objectref> annotations;
+```
+
+Here the addition of the annotations field could serve offer hints/guides to future passes (i.e. In Unified Static Memory Planner, we could use "candidate_memory_pools" as the key while the value being Map<String, PoolInfo>.)
+
+
+# Alternatives
+
+ ## S1. Using AttrStmt to associate additional info :
+
+TIR:
+ ```
+allocate_node_1 = tir.allocate([157323], "int16", "global")
+tir.attr(allocate_node_1, "pinned_memory", "foo_memory,bar_memory")
+ ``` 
+
+##  S2. Directly as an allocate node argument :
+
+ ```
+class AllocateNode : public StmtNode {
+ public:
+  /*! \brief The buffer variable. */
+  Var buffer_var;
+  /*! \brief The type of the buffer. */
+  DataType dtype;
+  /*! \brief The extents of the buffer. */
+  Array<PrimExpr> extents;
+  /*! \brief Only allocate buffer when condition is satisfied. */
+  PrimExpr condition;
+  /*! \brief The body to be executed. */
+  Stmt body;
+  /*! \brief If the allocate is scoped global, this field indicates
+   *  which external memories it could be pinned to as a comma seperated
+   *  string.
+   */
+  String pinned_memory;
+ ```
+TIR:
+ ```
+allocate_node_1 = tir.allocate([157323], "int16", "global",  "foo_memory,bar_memory")
+ ```
+
+ ## S3. Using additional tags in storage_scope
+
+```
+/*! \brief class to represent storage scope */
+struct StorageScope {
+  /*! \brief The rank of the storage */
+  StorageRank rank{StorageRank::kGlobal};
+  /*! \brief tag for special purpose memories. */
+  Array<String> tags;
+```
+
+
+ ```
+  /*!
+   * \brief Create storage scope from string
+   * \param s The string to be parsed.
+   * \return The storage scope.
+   */
+  static StorageScope Create(const std::string& s) {
+    StorageScope r;
+    if (s.empty()) {/
+      r.rank = StorageRank::kGlobal;
+    } else if (s.compare(0, 6, "global") == 0) {
+      r.rank = StorageRank::kGlobal;
+      r.tags = parseTags(s);
+ ```
+
+
+TIR:
+
+```
+allocate_node_1 = tir.allocate([157323], "int16", "global.(foo_memory,bar_memory)")
+```
+
+
+Out of the options, S1 seems the most non-invasive. However, we will need special handlers to obtain the information.
```

Comment:
however it does seem like this is a common problem with Attr and we could write a common one.

Reply:
I dont disagree but I feel that is out of scope for this RFC.

Reply:
Let's discuss this specific use in USMP RFC then. I do kind of agree that it's not in scope for an RFC that concerns itself only with adding the generic annotation/attr lookup to tir.allocate.